feat(animations): add support for disabling animations through BrowserAnimationsModule.withConfig

crisbeto opened the pull request:
Currently the only way to disable animations is by providing the `NoopAnimationsModule` which doesn't allow for it to be disabled based on runtime information. These changes add support for disabling animations based on runtime information by using `BrowserAnimationsModule.withConfig({disableAnimations: true})`.

JoostK commented on packages/platform-browser/animations/src/module.ts:
```diff
@@ -20,6 +29,27 @@ import {BROWSER_ANIMATIONS_PROVIDERS, BROWSER_NOOP_ANIMATIONS_PROVIDERS} from '.
+   * @param config Object used to configure the behavior of the `BrowserAnimationsModule`.
+   * @return The new `NgModule`.
+   */
+  static forRoot(config: BrowserAnimationsModuleConfig):
+      ModuleWithProviders<BrowserAnimationsModule|NoopAnimationsModule> {
```
I suppose this is problematic for the AOT compiler. The generic type of `ModuleWithProviders` has to be a simple type reference, as the compiler needs to statically know the NgModule import graph. The type union and dynamic switching as implemented here is therefore not supported.

The way I think this could work is to always return `BrowserAnimationsModule` here, but using the noop providers:

```ts
static forRoot(config: BrowserAnimationsModuleConfig): ModuleWithProviders<BrowserAnimationsModule> {
  return {
    ngModule: BrowserAnimationsModule,
    providers: config.disableAnimations ? BROWSER_NOOP_ANIMATIONS_PROVIDERS : BROWSER_ANIMATIONS_PROVIDERS,
  };
}
```

This removes the dynamic nature of the import while still providing the appropriate set of providers.

Additionally, it may be worth noting that even the above suggestion would likely _not_ work in View Engine, as it has an additional requirement that even the providers have to be statically known. This restriction has been lifted in Ivy. This can be partially worked around by capturing the `BrowserAnimationsModuleConfig` as a token of its own and then have `useFactory` providers for the other tokens that can then inject `BrowserAnimationsModuleConfig`. But even then you may not be able to do e.g. this:

```ts
BrowserAnimationsModule.forRoot({
  disableAnimations: window.matchMedia('(prefers-reduced-motion: reduce)').matches,
})
```

as the value for `disableAnimations` is not known statically, which I believe it is required to be in VE. You may want to experiment with this a bit, as I'm not entirely sure if there's a way to make it work in VE.

crisbeto replied:
I was suspecting it could be problematic, but shouldn't it have caused some of the unit tests to fail? 
And yeah, the dynamic value won't work for VE in AOT mode, but I figured that it won't be a big deal given that we're in the process of removing it.

JoostK replied:
In tests only the JIT compiler is used AFAIK. There's `packages/compiler-cli/integrationtest` but I have never worked with these, but it looks like the tests inside the `bazel` folder run the AOT compiler using both VE and Ivy.

AndrewKushnir reviewed (commented):
@crisbeto thanks for creating this PR 👍 

I have a couple questions/comments:
- should we consider calling the method `withConfig` instead of `forRoot` (since the root/child semantics from Router does not really apply here)
- Since it's possible to import the `BrowserAnimationsModule` for different sub-trees (and also call the `forRoot` now, such that one component might `disabledAnimations: true` and some other components might have `disableAnimations: false`), I'd propose to add some tests to verify that animations are properly enabled/disabled for the corresponding parts of the tree.

Thank you.

crisbeto commented:
I'm open to renaming `forRoot`, but I was under the impression that it's a general convention for methods like these. From the docs:
> The `forRoot()` static method is a convention that makes it easy for developers to configure services and providers that are intended to be singletons.

AndrewKushnir commented:
> I'm open to renaming forRoot, but I was under the impression that it's a general convention for methods like these.

Yeah, I think `forRoot` is more common in general, I'm just not sure this is what we want to express with this API (the `withConfig` sounds more inline with potential API usage). I've also found [`withConfig` usage](https://github.com/angular/angular/blob/master/packages/forms/src/form_providers.ts#L55-L64) in `ReactiveFormsModule` module.

May be we can get additional feedback from the `public-api` group reviewers on that (once this group is added to this PR)?

jelbourn commented:
This API can only be used at the root of the application though, right? If that's the case `forRoot` makes sense to me.

crisbeto commented:
Not necessarily. I believe that it would work on a subtree of modules as well.

jelbourn commented:
If that's the case then I think that `forRoot` wouldn't be the best fit since, at least in my mind, it's intended to only be used for APIs that only work at the app root NgModule

crisbeto commented:
Alright, I'll change it to `withConfig`.

crisbeto committed:
feat(animations): add support for disabling animations through BrowserAnimationsModule.withConfig

Currently the only way to disable animations is by providing the `NoopAnimationsModule`
which doesn't allow for it to be disabled based on runtime information. These changes
add support for disabling animations based on runtime information by using
`BrowserAnimationsModule.withConfig({disableAnimations: true})`.

crisbeto commented:
Changed based on the discussion @AndrewKushnir @jelbourn.

mary-poppins commented:
You can preview d11ede6 at https://pr40731-d11ede6.ngbuilds.io/.
You can preview 47c1f81 at https://pr40731-47c1f81.ngbuilds.io/.
You can preview e15f294 at https://pr40731-e15f294.ngbuilds.io/.
You can preview 9713bf9 at https://pr40731-9713bf9.ngbuilds.io/.
You can preview c570528 at https://pr40731-c570528.ngbuilds.io/.

AndrewKushnir commented on packages/core/test/animation/animation_integration_spec.ts:
```diff
@@ -56,9 +55,20 @@ describe('animation tests', function() {
+
+    it('should hint at NoopAnimationsModule being used when BrowserAnimationsModule is provided with disabled animations',
+       () => {
+         TestBed.resetTestingModule();
```
nit: the `TestBed.resetTestingModule` call should not be needed (here and in other tests), since this call is being executed between tests automatically.

crisbeto replied:
Removing it seems to mess up the test. I think it's because of the `beforeEach` earlier in the file.

AndrewKushnir commented on packages/platform-browser/animations/src/module.ts:
```diff
@@ -20,6 +29,29 @@ import {BROWSER_ANIMATIONS_PROVIDERS, BROWSER_NOOP_ANIMATIONS_PROVIDERS} from '.
 })
 export class BrowserAnimationsModule {
+  /**
+   * Creates and configures a module based on the specified object.
+   *
+   * When registering the NgModule at the root, import as follows:
```
```suggestion
   * Allows to configure the module by providing an object with the `BrowserAnimationsModuleConfig` shape.
   *
   * @see `BrowserAnimationsModuleConfig`
   *
   * @usageNotes
   * When registering the `BrowserAnimationsModule`, you can use the `withConfig` function as follows:
```

AndrewKushnir commented on packages/platform-browser/animations/src/module.ts:
````diff
@@ -20,6 +29,29 @@ import {BROWSER_ANIMATIONS_PROVIDERS, BROWSER_NOOP_ANIMATIONS_PROVIDERS} from '.
+   * class MyNgModule {}
+   * ```
+   *
+   * @param config Object used to configure the behavior of the `BrowserAnimationsModule`.
````
```suggestion
   * @param config Object with the `BrowserAnimationsModuleConfig` shape to configure the behavior of the module.
```

crisbeto replied:
Specifying what the shape of the object is supposed to be feels redundant since TS already enforces it.

AndrewKushnir replied:
Yeah, I was thinking to provide more specific info in docs, but I guess this would be derived from the function signature?

AndrewKushnir commented on packages/platform-browser/animations/src/module.ts:
````diff
@@ -20,6 +29,29 @@ import {BROWSER_ANIMATIONS_PROVIDERS, BROWSER_NOOP_ANIMATIONS_PROVIDERS} from '.
+   * ```
+   *
+   * @param config Object used to configure the behavior of the `BrowserAnimationsModule`.
+   * @return The new `NgModule`.
````
I guess we can just omit this line (it should be picked up by the docs engine)?

AndrewKushnir commented on packages/platform-browser/animations/src/module.ts:
```diff
@@ -5,11 +5,20 @@
+ * @publicApi
+ */
+export interface BrowserAnimationsModuleConfig {
+  /** Whether animations should be disabled. */
```
I'd propose to add a bit more description on what this config option does, what's the difference with the `NoopBrowserAnimationsModule` and when one should be used vs another.

AndrewKushnir commented on packages/core/test/animation/animation_integration_spec.ts:
```diff
@@ -56,9 +55,20 @@ describe('animation tests', function() {
+      expect(fixture.componentInstance.animationType).toEqual('NoopAnimations');
     });
+
+    it('should hint at NoopAnimationsModule being used when BrowserAnimationsModule is provided with disabled animations',
```
Can we add a test or two to verify the behavior when `BrowserAnimationsModule.withConfig` is used with different configs within the same components tree? Something like this:

```
@NgModule({
  declarations: [CompA],  // CompA.template = '<comp-b></comp-b>' (+ some animations)
  imports: [BrowserAnimationsModule.withConfig({disableAnimations: false})]
})
class ModuleA {}

@NgModule({
  declarations: [CompB], // CompB.template = '<comp-c></comp-c>' (+ some animations)
  imports: [BrowserAnimationsModule.withConfig({disableAnimations: true})]
})
class ModuleB {}

@NgModule({
  declarations: [CompC], // CompC.template = '...' (+ some animations)
  imports: [BrowserAnimationsModule.withConfig({disableAnimations: false})]
})
class ModuleC {}
```

crisbeto replied:
Wouldn't that be basically a test for the DI system? Also I'm not sure whether that would work as expected, because the modules would have to import each other which will end up picking the config from the top-most module (`ModuleA`).

AndrewKushnir replied:
Yeah, ok, thanks.

AndrewKushnir reviewed (commented):
Thanks for the updates @crisbeto! I've left a few comments.

crisbeto reviewed (commented):
I've addressed the latest set of feedback @AndrewKushnir.

mary-poppins commented:
You can preview 6f12954 at https://pr40731-6f12954.ngbuilds.io/.

crisbeto committed:
fixup! feat(animations): add support for disabling animations through BrowserAnimationsModule.withConfig

mary-poppins commented:
You can preview ae03b2a at https://pr40731-ae03b2a.ngbuilds.io/.

AndrewKushnir reviewed (approved):
LGTM, thanks for addressing the feedback @crisbeto 👍 

I wanted to have a quick look at how the `withConfig` function and the `BrowserAnimationsModuleConfig` interface would be presented on angular.io, but preview server seems to be down at the moment...

AndrewKushnir reviewed (approved):
Reviewed-for: public-api
Reviewed-for: size-tracking

AndrewKushnir commented:
[Presubmit](http://test/OCL:359322731:BASE:359322778:1614192849123:c051a69f).

jelbourn reviewed (approved):
LGTM. As a follow-up, we should update the animations guide to mention this new API and add an example. 

Reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: size-tracking
